README.md

# mapSpain <img src="man/figures/logo.png" align="right" width="120" />

<!-- badges: start -->

[![CRAN
status](https://www.r-pkg.org/badges/version/mapSpain)](https://CRAN.R-project.org/package=mapSpain)
[![CRAN
results](https://cranchecks.info/badges/worst/mapSpain)](https://cran.r-project.org/web/checks/check_results_mapSpain.html)
[![R-CMD-check](https://github.com/rOpenSpain/mapSpain/workflows/R-CMD-check/badge.svg)](https://github.com/rOpenSpain/mapSpain/actions?query=workflow%3AR-CMD-check)
[![codecov](https://codecov.io/gh/rOpenSpain/mapSpain/branch/master/graph/badge.svg?token=6L01BKLL85)](https://codecov.io/gh/rOpenSpain/mapSpain)
[![lifecycle](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://www.tidyverse.org/lifecycle/#experimental)
[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.4318024.svg)](https://doi.org/10.5281/zenodo.4318024)
<!-- badges: end -->

**[mapSpain](https://ropenspain.github.io/mapSpain/)** is a package that
provides spatial `sf` objects of the administrative boundaries of Spain,
including CCAA, provinces and municipalities.

**mapSpain** also provides a leaflet plugin to be used with the
[`leaflet` package](https://rstudio.github.io/leaflet/), that loads
several basemap of public institutions of Spain, and the ability of
downloading and processing static tiles.

Full site with examples and vignettes on
<https://ropenspain.github.io/mapSpain/>

## Installation

Install `mapSpain` from
[**CRAN**](https://CRAN.R-project.org/package=mapSpain):

``` r
install.packages("mapSpain")
```

You can install the developing version of `mapSpain` with:

``` r
library(remotes)
install_github("rOpenSpain/mapSpain")
```

## Usage

This script highlights some features of `mapSpain`:

``` r
library(mapSpain)
library(sf)
#> Linking to GEOS 3.5.1, GDAL 2.2.2, PROJ 4.9.2

census <- mapSpain::pobmun19

# Paste CCAA
census <-
  unique(merge(census, esp_codelist[, c("cpro", "codauto")], all.x = TRUE))

# Summarize
census_ccaa <-
  aggregate(cbind(pob19, men, women) ~ codauto, data = census, sum)

census_ccaa$porc_women <- census_ccaa$women / census_ccaa$pob19

CCAA.sf <- esp_get_ccaa()

# Merge data
CCAA.sf <- merge(CCAA.sf, census_ccaa)



# Choropleth map

library(cartography)
br <- getBreaks(CCAA.sf$porc_women, method = "pretty")
pal <- hcl.colors(length(br) - 1,
  palette = "Blues 3",
  rev = TRUE,
  alpha = 0.8
)
choroLayer(
  CCAA.sf,
  var = "porc_women",
  breaks = br,
  legend.values.rnd = 3,
  legend.pos = "bottomright",
  col = pal,
  border = NA,
  legend.title.txt = "% women"
)

# Add lines

CCAAbox <- esp_get_can_box()

plot(CCAAbox, add = TRUE, col = pal[length(pal)])
```

![](https://raw.githubusercontent.com/ropenspain/mapSpain/master/img/README-static-1.svg)<!-- -->

You can combine `POLYGONS` with static tiles

``` r
library(mapSpain)
library(sf)

census <- mapSpain::pobmun19
census$porc_women <- census$women / census$pob19

census$municode <- paste0(census$cpro, census$cmun)

Asturias.sf <- esp_get_munic(region = "Asturias")

Asturias.pop <-
  merge(Asturias.sf, census, by.x = "LAU_CODE", by.y = "municode")

# Get tiles

Asturias.pop <- st_transform(Asturias.pop, 3857)

tile <-
  esp_getTiles(Asturias.pop,
    type = "IGNBase.Todo",
    zoom = 8
  )


# Plot
par(mar = c(0, 0, 0, 0))
library(cartography)

tilesLayer(tile)
br <- getBreaks(Asturias.pop$porc_women, method = "kmeans")


choroLayer(
  Asturias.pop,
  var = "porc_women",
  breaks = br,
  col = hcl.colors(
    n = length(br) - 1,
    palette = "inferno",
    rev = TRUE,
    alpha = 0.4
  ),
  add = TRUE,
  legend.values.rnd = 2,
  border = NA,
  legend.pos = "n"
)

legendChoro(
  pos = "bottomright",
  breaks = br,
  values.rnd = 2,
  horiz = TRUE,
  nodata = FALSE,
  col = hcl.colors(
    n = length(br) - 1,
    palette = "inferno",
    rev = TRUE
  ),
  title.txt = "% women"
)
```

![](https://raw.githubusercontent.com/ropenspain/mapSpain/master/img/README-tile-1.svg)<!-- -->

### A note on caching

Some data sets and tiles may have a size larger than 50MB. You can use
`mapSpain` to create your own local repository at a given local
directory passing the following option:

``` r
options(mapSpain_cache_dir = "./path/to/location")
```

When this option is set, `mapSpain` would look for the cached file and
it will load it, speeding up the process.

### Plotting `sf` objects

Some packages recommended for visualization are:

  - [`tmap`](https://mtennekes.github.io/tmap/)  
  - [`cartography`](http://riatelab.github.io/cartography/docs/)
  - [`ggplot2`](https://github.com/tidyverse/ggplot2) +
    [`ggspatial`](https://github.com/paleolimbot/ggspatial)
  - [`leaflet`](https://rstudio.github.io/leaflet/)

## Contribute

Check the Github page for [source
code](https://github.com/ropenspain/mapSpain/).

## Copyright notice

This package uses data from **GISCO**. GISCO
[(FAQ)](https://ec.europa.eu/eurostat/web/gisco/faq) is a geospatial
open data repository including several data sets at several resolution
levels.

*From GISCO \> Geodata \> Reference data \> Administrative Units /
Statistical Units*

When data downloaded from this page is used in any printed or electronic
publication, in addition to any other provisions applicable to the whole
Eurostat website, data source will have to be acknowledged in the legend
of the map and in the introductory page of the publication with the
following copyright notice:

EN: © EuroGeographics for the administrative boundaries

FR: © EuroGeographics pour les limites administratives

DE: © EuroGeographics bezüglich der Verwaltungsgrenzen

For publications in languages other than English, French or German, the
translation of the copyright notice in the language of the publication
shall be used.

If you intend to use the data commercially, please contact
EuroGeographics for information regarding their license agreements.


NEWS.md
# mapSpain 0.1.3-9000

* Fix DOI <https://doi.org/10.5281/zenodo.4318024>
* WIP: Include CartoBase ANE data <https://github.com/rOpenSpain/mapSpain/tree/sianedata>
  * `mapSpain::esp_get_munic_siane()`



# mapSpain 0.1.2

* Fix annoying warning if `sf` was not loaded first.
* Include new `poly` option on `mapSpain::esp_get_can_box()`.
* New grids created with `geogrid::calculate_grid()`
* Add more years on `mapSpain::esp_get_munic()`
* Move to rOpenSpain organization.

# mapSpain 0.1.1

* Fix CRAN submission.
* Added `mapSpain::esp_get_grid_prov()` and `mapSpain::esp_get_grid_ccaa()`.



# mapSpain 0.1.0

* Initial release.


inst/tinytest/test_esp_get_ccaa.R
library(tinytest)

expect_error(esp_get_ccaa("FFF"))
expect_silent(esp_get_ccaa())
expect_silent(esp_get_ccaa(ccaa = c("Galicia", "ES7", "Centro")))
expect_warning(esp_get_ccaa(ccaa = "Zamora"))
expect_warning(esp_get_ccaa(ccaa = "ES6x"))
expect_warning(esp_get_ccaa(ccaa = "Barcelona"))


# Test all

f <- mapSpain::esp_codelist

n <- esp_get_ccaa(ccaa = f$nuts1.code)
expect_equal(nrow(n), 19)

n <- esp_get_ccaa(ccaa = c("Melilla"))
expect_equal(nrow(n), 1)

n <- esp_get_ccaa(ccaa = f$nuts1.name.alt)
expect_equal(nrow(n), 19)

n <- esp_get_ccaa(ccaa = f$iso2.ccaa.code)
expect_equal(nrow(n), 19)

n <- esp_get_ccaa(ccaa = f$nuts2.code)
expect_equal(nrow(n), 19)

n <- esp_get_ccaa(ccaa = f$nuts2.name)
expect_equal(nrow(n), 19)

n <- esp_get_ccaa(ccaa = f$codauto)
expect_equal(nrow(n), 19)


# Test siane


expect_error(esp_get_ccaa_siane("FFF"))
expect_error(esp_get_ccaa_siane(epsg = "FFF"))
expect_silent(esp_get_ccaa_siane())
expect_silent(esp_get_ccaa_siane(ccaa = c("Galicia", "ES7", "Centro")))
expect_error(esp_get_ccaa_siane(epsg = 39823))
expect_silent(esp_get_ccaa_siane())
expect_silent(esp_get_ccaa_siane(moveCAN = c(1, 2)))
expect_silent(esp_get_ccaa_siane(ccaa = c("Galicia", "ES7", "Centro")))
expect_warning(esp_get_ccaa_siane(ccaa = "Menorca"))
expect_warning(esp_get_ccaa_siane(ccaa = "ES6x"))


expect_equal(sf::st_crs(esp_get_ccaa_siane(epsg = 3035)),
             sf::st_crs(3035))

expect_equal(sf::st_crs(esp_get_ccaa_siane(epsg = 3857)),
             sf::st_crs(3857))



# Test all

f <- mapSpain::esp_codelist

n <- esp_get_ccaa_siane(ccaa = f$nuts1.code)
expect_equal(nrow(n), 19)

n <- esp_get_ccaa_siane(ccaa = c("Melilla"))
expect_equal(nrow(n), 1)

n <- esp_get_ccaa_siane(ccaa = f$nuts1.name.alt)
expect_equal(nrow(n), 19)

n <- esp_get_ccaa_siane(ccaa = f$iso2.ccaa.code)
expect_equal(nrow(n), 19)

n <- esp_get_ccaa_siane(ccaa = f$nuts2.code)
expect_equal(nrow(n), 19)

n <- esp_get_ccaa_siane(ccaa = f$nuts2.name)
expect_equal(nrow(n), 19)

n <- esp_get_ccaa_siane(ccaa = f$codauto)
expect_equal(nrow(n), 19)


inst/tinytest/test_esp_get_prov.R
library(tinytest)

expect_error(esp_get_prov("FFF"))
expect_silent(esp_get_prov())
expect_silent(esp_get_prov(prov = c("Galicia", "ES7", "Centro")))
expect_warning(esp_get_prov(prov = "Menorca"))
expect_warning(esp_get_prov(prov = "ES6x"))

expect_silent(esp_get_prov(prov = c(
  "Euskadi",
  "Catalunya",
  "ES-EX",
  "ES52",
  "01"
)))

# Test all

f <- mapSpain::esp_codelist

n <- esp_get_prov(prov = f$nuts1.code)
expect_equal(nrow(n), 52)

n <- esp_get_prov(prov = c("Canarias"))
expect_equal(nrow(n), 2)

n <- esp_get_prov(prov = c("Baleares"))
expect_equal(nrow(n), 1)

n <- esp_get_prov(prov = f$nuts1.name.alt)
expect_equal(nrow(n), 52)

n <- esp_get_prov(prov = f$iso2.prov.code)
expect_equal(nrow(n), 52)

n <- esp_get_prov(prov = f$nuts2.code)
expect_equal(nrow(n), 52)

n <- esp_get_prov(prov = f$nuts2.name)
expect_equal(nrow(n), 52)

n <- esp_get_prov(prov = f$cpro)
expect_equal(nrow(n), 52)

n <- esp_get_prov(prov = f$prov.shortname.es)
expect_equal(nrow(n), 52)

# SIANE

expect_error(esp_get_prov_siane("FFF"))
expect_error(esp_get_prov_siane(epsg = 39823))
expect_silent(esp_get_prov_siane())
expect_silent(esp_get_prov_siane(moveCAN = c(1, 2)))
expect_silent(esp_get_prov_siane(prov = c("Galicia", "ES7", "Centro")))
expect_warning(esp_get_prov_siane(prov = "Menorca"))
expect_warning(esp_get_prov_siane(prov = "ES6x"))


expect_equal(sf::st_crs(esp_get_prov_siane(epsg = 3035)),
             sf::st_crs(3035))

expect_equal(sf::st_crs(esp_get_prov_siane(epsg = 3857)),
             sf::st_crs(3857))


expect_silent(esp_get_prov_siane(prov = c(
  "Euskadi",
  "Catalunya",
  "ES-EX",
  "ES52",
  "01"
)))

# Test all

f <- mapSpain::esp_codelist

n <- esp_get_prov_siane(prov = f$nuts1.code)
expect_equal(nrow(n), 52)

n <- esp_get_prov_siane(prov = c("Canarias"))
expect_equal(nrow(n), 2)

n <- esp_get_prov_siane(prov = c("Baleares"))
expect_equal(nrow(n), 1)

n <- esp_get_prov_siane(prov = f$nuts1.name.alt)
expect_equal(nrow(n), 52)

n <- esp_get_prov_siane(prov = f$iso2.prov.code)
expect_equal(nrow(n), 52)

n <- esp_get_prov_siane(prov = f$nuts2.code)
expect_equal(nrow(n), 52)

n <- esp_get_prov_siane(prov = f$nuts2.name)
expect_equal(nrow(n), 52)

n <- esp_get_prov_siane(prov = f$cpro)
expect_equal(nrow(n), 52)

n <- esp_get_prov_siane(prov = f$prov.shortname.es)
expect_equal(nrow(n), 52)


inst/tinytest/test_esp_get_munic.R
library(tinytest)

expect_silent(esp_get_munic())
expect_silent(esp_get_munic(moveCAN = FALSE))
expect_silent(esp_get_munic(moveCAN = c(0, 10)))
expect_error(esp_get_munic(region = "XX"))
expect_error(esp_get_munic(year = "2040"))
expect_error(esp_get_munic(munic = "XX"))
expect_silent(esp_get_munic(munic = "Nieva"))
expect_silent(esp_get_munic(region = "Alava"))
expect_message(esp_get_munic(region = "Canarias", verbose = TRUE))

a <- mapSpain::esp_codelist
n <- a$nuts1.name

s <- esp_get_munic(region = n)
expect_equal(length(unique(s$cpro)), 52)


# SIANE

expect_silent(esp_get_munic_siane())

expect_message(esp_get_munic_siane(verbose = TRUE))
expect_message(esp_get_munic_siane(verbose = TRUE, update_cache = TRUE))
expect_message(esp_get_munic_siane(region = "Canarias", verbose = TRUE))

expect_error(esp_get_munic_siane(year = "2019-15-23"))
expect_error(esp_get_munic_siane(year = "2019-15"))
expect_error(esp_get_munic_siane(region = "XX"))
expect_error(esp_get_munic_siane(epsg = "5689"))
expect_error(esp_get_munic_siane(resolution = 5.6))
expect_error(esp_get_munic_siane(year = "2040"))
expect_error(esp_get_munic_siane(munic = "XX"))

expect_silent(esp_get_munic_siane(moveCAN = FALSE))
expect_silent(esp_get_munic_siane(moveCAN = c(0, 10)))
expect_silent(esp_get_munic_siane(year = "2019-10-23"))
expect_silent(esp_get_munic_siane(munic = "Nieva"))
expect_silent(esp_get_munic_siane(region = "Alava"))


a <- mapSpain::esp_codelist
n <- a$nuts1.name

s <- esp_get_munic_siane(region = n)
expect_equal(length(unique(s$cpro)), 52)


dev/dev.R
roxygen2::roxygenise()

remotes::install_github("rOpenSpain/mapSpain")

tinytest::test_all()


a <- esp_get_ccaa()

f <- esp_get_hex_ccaa()
plot(st_geometry(f))

f <- esp_get_hex_ccaa()
f <- st_make_valid(f)
st_is_valid(f)

st_crs(f)

library(mapSpain)

s <- esp_get_hex_prov()
s <- st_transform(s,3857)
st_is_valid(s)
par(mar=c(0,0,0,0))
plot(st_geometry(s))

devtools::build_readme()

tinytest::test_all()

help(giscoR)

usethis::use_cran_comments()

devtools::spell_check()
codemetar::write_codemeta(write_minimeta = TRUE)

??mapSpain

devtools::release()

devtools::check()

tinytest::test_all()

options(mapSpain_cache_dir="~/R/mapslib/GISCO")

devtools::revdep("mapSpain")
pkgdown::build_articles()
pkgdown::clean_site()
pkgdown::build_article("mapSpain")

pkgdown::build_articles_index()

pkgdown::build_site(lazy = TRUE)

devtools::spell_check()

pkgdown::clean_site()

sessionInfo()

tinytest::test_all()

hcl.pals("sequential")

devtools::check()
devtools::check_win_release()
devtools::check_win_devel()
devtools::check_win_oldrelease()
devtools::check_rhub()

dev

codemetar::write_codemeta(write_minimeta = TRUE)


devtools::build_manual(path = "./dev")
pkgdown::build_favicons(overwrite = TRUE)

pkgdown::build_site(lazy = TRUE)
pkgdown::build_articles(lazy = FALSE)

pkgdown::build_site()
pkgdown::build_reference_index()

pkgdown::build_favicons(overwrite = TRUE)

usethis::use_readme_rmd()
devtools::build_readme()
pkgdown::build_()

pkgdown::clean_site()
devtools::document()
interactive()

leaflet.providersESP.df

leafle
tinytest::test_all()



# Checks release


devtools::spell_check()
devtools::check()
devtools::check_rhub()
devtools::check_win_release()
devtools::check_win_devel()
devtools::check_win_oldrelease()

devtools::release()

install.packages("devtools",dependencies = TRUE)

rhub::validate_email()

leaflet::addP

point <- st_point(c(-3.6886664,40.3922413))
point <- st_sfc(point)
point <- st_set_crs(point,4326)
point <- st_transform(point,3857)
#point <- st_buffer(point,1000)
a <- esp_getTiles(point, type = "Catastro.Parcela",
                           verbose = TRUE,
                  crop = TRUE,
                  res=256)


raster::plotRGB(a)
plot(st_geometry(point), add=TRUE)

point

library(mapSpain)

sessionInfo()

f <- mapSpain::leaflet.providersESP
f <- unique(f$provider)


library(goodpractice)

gp()

tinytest::test_all()

xx <- esp_get_ccaa()

xx <- sf::st_transform(xx,3857)

options(mapSpain_cache_dir = "~/R/mapslib/GISCO")

rt <- esp_getTiles(xx, type = f[1], verbose = TRUE, zoom = 4 )

par(mar=c(0,0,0,0))
cartography::tilesLayer(rt)

lintr::lint_package()


tinytest::test_all()
covr::report()


ff <- mapSpain::leaflet.providersESP

leaflet::addp

addTiles(
  map,
  urlTemplate = "//{s}.tile.openstreetmap.org/{z}/{x}/{y}.png",
  attribution = NULL,
  layerId = NULL,
  group = NULL,
  options = tileOptions(),
  data = getMapData(map)
)


addProviderTiles(
  map,
  provider,
  layerId = NULL,
  group = NULL,
  options = providerTileOptions()
)

providerTileOptions(
  errorTileUrl = "",
  noWrap = FALSE,
  opacity = NULL,
  zIndex = NULL,
  updateWhenIdle = NULL,
  detectRetina = FALSE,
  ...
)

aa <- leaflet::WMSTileOptions()
bb <- leaflet::tileOptions(arg = 1)


ff <- mapSpain::leaflet.providersESP

library(leaflet)

cc <- do.call( leaflet::tileOptions,c(aa,bb))

cc

WMSTileOptions(
  styles = "",
  format = "image/jpeg",
  transparent = FALSE,
  version = "1.1.1",
  crs = NULL,
  ...
)

addWMSTiles(
  map,
  baseUrl,
  layerId = NULL,
  group = NULL,
  options = WMSTileOptions(),
  attribution = NULL,
  layers = "",
  data = getMapData(map)
)


devtools::build_manual("./dev")

library(goodpractice)

gp()

lintr::lint_package()

f <- esp_get_nuts(region = "Madrid", moveCAN = TRUE, epsg = 3857)
p <- esp_getTiles(f, type = "IGNBase.Gris", verbose = TRUE)
par(mar=c(0,0,0,0))
cartography::tilesLayer(p)


x <- f
update_cache = FALSE
cache_dir = NULL
verbose = FALSE
type = "IDErioja"
zoom = NULL
crop = TRUE
res = 512



xinit <- x

if (nrow(x) == 1 && sf::st_is(x, "POINT")) {
  x <- sf::st_transform(x, 3857)
  x <- sf::st_buffer(sf::st_geometry(x), 1000)
  crop <- FALSE
}

leafletProvidersESP <- mapSpain::leaflet.providersESP
provs <-
  leafletProvidersESP[leafletProvidersESP$provider == type,]

if (nrow(provs) == 0) {
  stop(
    "No match for type = '",
    type,
    "' found. Available providers are:\n\n",
    paste0("'", unique(leafletProvidersESP$provider), "'", collapse = ", ")
  )

}


cache_dir <- esp_hlp_cachedir(cache_dir)
cache_dir <- esp_hlp_cachedir(paste0(cache_dir, "/", type))


x <- sf::st_transform(x, 4326)
bbx <-
  sf::st_bbox(sf::st_transform(sf::st_as_sfc(sf::st_bbox(x)), 4326))

# select a default zoom level
if (is.null(zoom)) {
  gz <- slippymath::bbox_tile_query(bbx)
  zoom <- min(gz[gz$total_tiles %in% 4:10, "zoom"])

  if (verbose) {
    message("Auto zoom level: ", zoom)
  }
}
minZoom <- as.numeric()

!is.na(as.numeric(minZoom))

if (length(minZoom) > 0) {
  zoom <- max(zoom, minZoom)

  if (verbose & minZoom == zoom)
    message("\nSwiching. Minimum zoom for this provider is ", zoom, "\n")


}



R/esp_get_prov.R
#' @title Get Provinces boundaries of Spain
#' @name esp_get_prov
#' @description Loads a simple feature (\code{sf}) object containing the
#' province boundaries of Spain.
#'
#' \code{esp_get_prov} uses GISCO (Eurostat) as source
#'
#' @return A \code{POLYGON/POINT} object.
#' @author dieghernan, \url{https://github.com/dieghernan/}
#' @seealso \link{esp_get_hex_prov}, \link{esp_get_nuts}, \link{esp_get_ccaa},
#' \link{esp_get_munic}, \link{esp_codelist}
#' @export
#'
#'
#'
#' @param prov A vector of names and/or codes for provinces
#'  or \code{NULL} to get all the provinces. See Details.
#' @param ... Additional parameters from \link{esp_get_nuts}.
#' @details When using \code{prov} you can use and mix names and NUTS codes
#' (levels 1, 2 or 3), ISO codes (corresponding to level 2 or 3) or
#' \code{cpro}.
#'
#' Ceuta and Melilla are considered as provinces on this dataset.
#'
#' When calling a superior level (Autonomous Community or NUTS1) ,
#' all the provinces of that level would be added.
#' @examples
#'
#' library(sf)
#'
#' # Random Provinces
#'
#' Random <-
#'   esp_get_prov(prov = c("Zamora",
#'                         "Palencia",
#'                         "ES-GR",
#'                         "ES521",
#'                         "01"))
#' plot(st_geometry(Random), col = hcl.colors(6))
#'
#' # All Provinces of a Zone plus an addition
#'
#' Mix <-
#'   esp_get_prov(prov = c("Noroeste",
#'                         "Castilla y Leon", "La Rioja"),
#'                resolution = "20")
#' plot(
#'   Mix[, "nuts1.code"],
#'   pal = hcl.colors(3),
#'   key.pos = NULL,
#'   main = NULL,
#'   border = "white"
#' )

esp_get_prov <- function(prov = NULL, ...) {
  params <- list(...)

  # Get region id

  prov <- prov[!is.na(prov)]


  region <- prov
  if (is.null(region)) {
    nuts_id <- NULL
  } else {
    nuts_id <- esp_hlp_all2prov(region)

    nuts_id <- unique(nuts_id)
    if (length(nuts_id) == 0)
      stop("region ",
           paste0("'", region, "'", collapse = ", "),
           " is not a valid name")
  }

  params$region <- nuts_id
  params$nuts_level <- 3


  data.sf <- do.call(mapSpain::esp_get_nuts,  params)

  data.sf$nuts3.code <- data.sf$NUTS_ID
  data.sf <- data.sf[, "nuts3.code"]

  # Get cpro

  df <- mapSpain::esp_codelist
  df <- unique(df[, c("nuts3.code", "cpro")])
  data.sf <- merge(data.sf, df, all.x = TRUE)
  data.sf <- data.sf[, "cpro"]

  # Merge Islands

  # Extract geom column
  names <- names(data.sf)

  which.geom <-
    which(vapply(data.sf, function(f)
      inherits(f, "sfc"), TRUE))

  nm <- names(which.geom)

  # Baleares

  if (any(data.sf$cpro %in% "07")) {
    clean <- data.sf[data.sf$cpro != "07", ]
    island <- data.sf[data.sf$cpro == "07", ]
    g <- sf::st_union(island)
    df <- unique(sf::st_drop_geometry(island))
    # Generate sf object
    new <- sf::st_as_sf(df, g)
    # Rename geometry to original value
    newnames <- names(new)
    newnames[newnames == "g"] <- nm
    colnames(new) <- newnames
    new <- sf::st_set_geometry(new, nm)
    data.sf <- rbind(clean, new)
  }

  # Las Palmas

  if (any(data.sf$cpro %in% "35")) {
    clean <- data.sf[data.sf$cpro != "35", ]
    island <- data.sf[data.sf$cpro == "35", ]
    g <- sf::st_union(island)
    df <- unique(sf::st_drop_geometry(island))
    # Generate sf object
    new <- sf::st_as_sf(df, g)
    # Rename geometry to original value
    newnames <- names(new)
    newnames[newnames == "g"] <- nm
    colnames(new) <- newnames
    new <- sf::st_set_geometry(new, nm)
    data.sf <- rbind(clean, new)
  }

  # Sta Cruz Tfe
  if (any(data.sf$cpro %in% "38")) {
    clean <- data.sf[data.sf$cpro != "38", ]
    island <- data.sf[data.sf$cpro == "38", ]
    g <- sf::st_union(island)
    df <- unique(sf::st_drop_geometry(island))
    # Generate sf object
    new <- sf::st_as_sf(df, g)
    # Rename geometry to original value
    newnames <- names(new)
    newnames[newnames == "g"] <- nm
    colnames(new) <- newnames
    new <- sf::st_set_geometry(new, nm)
    data.sf <- rbind(clean, new)
  }



  # Get df
  df <- dict_prov
  df <- df[, names(df) != "key"]

  data.sf <- merge(data.sf, df, all.x = TRUE)

  # Paste nuts2
  dfnuts <- mapSpain::esp_codelist
  dfnuts <-
    unique(dfnuts[, c("cpro",
                      "nuts2.code",
                      "nuts2.name",
                      "nuts1.code",
                      "nuts1.name")])
  data.sf <- merge(data.sf, dfnuts, all.x = TRUE)
  data.sf <-
    data.sf[, c(colnames(df),
                "nuts2.code",
                "nuts2.name",
                "nuts1.code",
                "nuts1.name")]

  # Order
  data.sf <- data.sf[order(data.sf$codauto, data.sf$cpro), ]

  return(data.sf)
}


#' @rdname esp_get_prov
#' @description \code{esp_get_prov_siane} use CartoBase ANE as source,
#' provided by Instituto Geografico Nacional (IGN),
#' \href{http://www.ign.es/web/ign/portal}{ign.es}. Years available are
#' 2005 up to today.
#' @source IGN data via a custom CDN (see
#' \url{https://github.com/rOpenSpain/mapSpain/tree/sianedata}).
#' @export
#'
#'
#' @param resolution Resolution of the polygon. Values available are
#' \code{"3", "6.5"} or  \code{"10"}.
#' @param rawcols Logical. Setting this to \code{TRUE} would add the raw
#' columns of the dataset provided by IGN.
#' @param year,epsg,cache,update_cache,cache_dir,verbose,moveCAN
#' See \code{\link{esp_get_nuts}}. See Details for \code{year}.
#'
#' @details
#' On \code{esp_get_prov_siane}, \code{year} could be passed as a single
#' year ("YYYY" format, as end of year) or as a specific
#' date ("YYYY-MM-DD" format). Historical information starts as of 2005.

esp_get_prov_siane <- function(prov = NULL,
                               year = Sys.Date(),
                               epsg = "4258",
                               cache = TRUE,
                               update_cache = FALSE,
                               cache_dir = NULL,
                               verbose = FALSE,
                               resolution = "3",
                               moveCAN = TRUE,
                               rawcols = FALSE) {
  init_epsg <- as.character(epsg)
  year <- as.character(year)


  if (!init_epsg %in% c("4326", "4258", "3035", "3857")) {
    stop("epsg value not valid. It should be one of 4326, 4258, 3035 or 3857")
  }

  # Get Data from SIANE
  data.sf <- esp_hlp_get_siane("prov",
                               resolution,
                               cache,
                               cache_dir,
                               update_cache,
                               verbose,
                               year)


  colnames_init <- colnames(sf::st_drop_geometry(data.sf))


  data.sf$cpro <- data.sf$id_prov

  if (!is.null(prov)) {
    tonuts <- esp_hlp_all2prov(prov)
    #toprov
    df <- unique(mapSpain::esp_codelist[, c("nuts3.code", "cpro")])
    df <- df[df$nuts3.code %in% tonuts, "cpro"]
    toprov <- unique(df)
    data.sf <- data.sf[data.sf$cpro %in% toprov, ]
  }

  if (nrow(data.sf) == 0) {
    stop("provs '",
         paste0(prov, collapse = "', "),
         "' does not return any result")
  }

  # Get df
  df <- dict_prov
  df <- df[, names(df) != "key"]

  data.sf <- merge(data.sf, df, all.x = TRUE)

  # Paste nuts2
  dfnuts <- mapSpain::esp_codelist
  dfnuts <-
    unique(dfnuts[, c("cpro",
                      "nuts2.code",
                      "nuts2.name",
                      "nuts1.code",
                      "nuts1.name")])
  data.sf <- merge(data.sf, dfnuts, all.x = TRUE)


  # Move CAN

  # Checks
  moving <- FALSE
  moving <- isTRUE(moveCAN) | length(moveCAN) > 1

  if (moving) {
    if (length(grep("05", data.sf$codauto)) > 0) {
      offset <- c(550000, 920000)

      if (length(moveCAN) > 1) {
        coords <- sf::st_point(moveCAN)
        coords <- sf::st_sfc(coords, crs = sf::st_crs(4326))
        coords <- sf::st_transform(coords, 3857)
        coords <- sf::st_coordinates(coords)
        offset <- offset + as.double(coords)
      }

      data.sf <- sf::st_transform(data.sf, 3857)
      PENIN <- data.sf[-grep("05", data.sf$codauto), ]
      CAN <- data.sf[grep("05", data.sf$codauto), ]

      # Move CAN
      CAN <- sf::st_sf(
        sf::st_drop_geometry(CAN),
        geometry = sf::st_geometry(CAN) + offset,
        crs = sf::st_crs(CAN)
      )

      #Regenerate
      if (nrow(PENIN) > 0) {
        data.sf <- rbind(PENIN, CAN)
      } else {
        data.sf <- CAN
      }
    }
  }

  data.sf <- sf::st_transform(data.sf, as.double(init_epsg))


  # Order
  data.sf <- data.sf[order(data.sf$codauto, data.sf$cpro), ]

  namesend <- unique(c(colnames_init,
                       colnames(esp_get_prov())))

  data.sf <- data.sf[, namesend]


  if (isFALSE(rawcols)) {
    data.sf <- data.sf[, colnames(esp_get_prov())]
  }

  return(data.sf)
}
